Add support for dml_fail_fast session setting

BaurzhanSakhariev opened the pull request:
Supersedes https://github.com/crate/crate/pull/15582

Resolves https://github.com/crate/crate/issues/12218

TODO:

- [ ] Update executeBulk methods i tests to accept session (so that we don't have to use PostresIT)
- [ ] Caught exceptions in tests are semantically correct but not client friendly.
Missing `prepareForClientTransmission` somewhere along the line.
- [ ] Docs

BaurzhanSakhariev commented:
Hi @mfussenegger, can you please provide an early feedback on this draft?

Approach: use existing `continueOnError` flag to stop an operation on a shard.

Details:

1.  COPY FROM is not affected, see comment in the `IndexWriteProjector`, it has its own fail_fast 
2. DELETE and bulk DeleteById are not affected (because they can't really fail). Maybe need to rename the setting from dml to upsert then?
3. https://github.com/crate/crate/pull/15582 had a flaw that Session was setting aware. In this PR only upsert components are aware about the setting. 


Why it's draft:

As far as I know, we need to send KILL requests to stop the operation across all shards(?)

Comment from @jeeminso 
>I think ultimately earlyTerminationCondition need to be set in order to stop the entire upserts across all shards.
If only  request.continueOnError() is flipped these is what happens:
```
cr> create table t (a int check(a != 10));
CREATE OK, 1 row affected (1.713 sec)
cr> insert into t select * from generate_series(1, 100000);
INSERT OK, 97432 rows affected (1.035 sec) -- over 90% inserted, doesn't seem right
```

I will adapt this PR to integrate fail_fast mechanism used in COPY FROM that was introduced in https://github.com/crate/crate/pull/11607

BaurzhanSakhariev commented:
>As far as I know, we need to send KILL requests to stop the operation across all shards(?)

This is handled by `InterceptingRowConsumer`, thanks @seut for pointing it out!

mfussenegger commented on docs/config/session.rst:
```diff
@@ -276,4 +276,13 @@ consumption for the particular query.
 
+.. _conf-session-dml-fail-fast:
+
+**dml_fail_fast**
+  | *Default:* ``false``
+  | *Modifiable:* ``yes``
+   This setting enables partial failures of ``INSERT`` and ``UPDATE``
+   statements. If enabled, first error encountered gets propagated, operation
+   is terminated and number of inserted records is not reported.
```
I'm not sure about the naming. I'd tend to highlight the partial write - e.g. `allow_fail_on_partial_writes`

BaurzhanSakhariev replied:
renamed in https://github.com/crate/crate/pull/17689/commits/41882d43fc9dd4773156b8d2c8a96f36ef0a5ebe

mfussenegger commented on server/src/main/java/io/crate/planner/operators/InsertFromValues.java:
```diff
@@ -453,10 +465,16 @@ public CompletableFuture<BulkResponse> executeBulk(DependencyCarrier dependencie
         }).whenComplete((response, t) -> {
             if (t == null) {
                 bulkResponse.update(response, bulkIndices);
-                result.complete(bulkResponse);
             } else {
-                result.completeExceptionally(t);
+                assert response == null : "No response when one bulk short circuited." +
+                    "Creating a bulkResponse by setting up the same error to all bulks.";
+                for (IntCursor c : bulkIndices) {
+                    int resultIdx = c.value;
+                    t = SQLExceptions.prepareForClientTransmission(accessControl, t);
+                    bulkResponse.setFailure(resultIdx, t);
+                }
             }
+            result.complete(bulkResponse);
```
I don't understand why we're changing the bulk response handling too?

It already has bulk parameter dependend `Throwable[] failures` - all we need to do is to ensure the throwable/failures are set, then the reporting to the client (and here) should all already fit. Or what am I missing?

BaurzhanSakhariev replied:
It all boils down to exceptions being swallowed on emitting final result, ie we don't have 
`exceptionally` in addition to
` .thenAccept(bulkResp -> emitRowCountsToResultReceivers(jobId, jobsLogs, toExec, bulkResp));`

The thing is  `result.completeExceptionally(t);` was never hit on master
(because we never had `continueOnError` flipped)

If I revert this, `test_bulk_insert_from_values_fail_fast` gets stuck. Then it kinda works if I add `exceptionally` to the Session (to catch bubbled up exception) but there we have only error, no `BulkResponse`, so we don't know which ResultResiver we should fail and which one just finish.

BaurzhanSakhariev replied:
bulkresponse.failures contains 2 nulls BTW, we need to enrich it by the actual failure we caught.

When future completes with a proper response, we can build response properly via  `bulkResponse.update(response, bulkIndices);` when we have only error, we build response based on the failure

seut commented on server/src/test/java/io/crate/integrationtests/UpdateIntegrationTest.java:
```diff
@@ -1134,4 +1134,44 @@ public void test_update_continues_on_error() {
+    }
+
+    @Test
+    public void test_bulk_update_by_id_fail_fast() throws Exception {
```
`UpdateById` must also taken the new session setting into account. Currently, the operation won't stop, but just the first error is returned.
Please adjust the test, so that it would write a valid record after the failure and then ensure that his record was never written. Same for the UpdateByQuery test.

seut commented on server/src/test/java/io/crate/metadata/settings/SessionSettingsTest.java:
```diff
@@ -36,7 +36,7 @@ public class SessionSettingsTest {
     @Test
     public void testSessionSettingsStreaming() throws IOException {
-        SessionSettings s1 = new SessionSettings("user", SearchPath.createSearchPathFrom("crate"), true, false, 20);
+        SessionSettings s1 = new SessionSettings("user", SearchPath.createSearchPathFrom("crate"), true, false, 20, false);
```
`false` is the default value, so streaming the correct value is only ensured when using `true`.
```suggestion
        SessionSettings s1 = new SessionSettings("user", SearchPath.createSearchPathFrom("crate"), true, false, 20, true);
```

BaurzhanSakhariev commented on server/src/main/java/io/crate/planner/operators/InsertFromValues.java:
```diff
@@ -453,10 +465,16 @@ public CompletableFuture<BulkResponse> executeBulk(DependencyCarrier dependencie
-                result.complete(bulkResponse);
             } else {
-                result.completeExceptionally(t);
+                assert response == null : "No response when one bulk short circuited." +
+                    "Creating a bulkResponse by setting up the same error to all bulks.";
+                for (IntCursor c : bulkIndices) {
+                    int resultIdx = c.value;
+                    t = SQLExceptions.prepareForClientTransmission(accessControl, t);
+                    bulkResponse.setFailure(resultIdx, t);
+                }
```
See https://github.com/crate/crate/pull/17689#discussion_r2022483529

BaurzhanSakhariev replied:
Yep, I'm switching my approach to add `exceptionally`.

BaurzhanSakhariev replied:
>that this was missing was wrong in the first place I think (existing bug?). E.g.:

I think was done by intention, we even had a code comment that we removed recently
https://github.com/crate/crate/commit/c7fb24dd00e0f14c8a2ab6f9c99ef5c663e85e8b#diff-35c236fb9bdd18dcf79b53411d3e8b235bb9ddb95193e439aaf35eb0fb9295e7L707

>   .exceptionally(t -> null) // swallow exception - failures are set per item in emitResults

I think we never threw an error as we never had `continueOnError` flipped

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/dml/BulkResponse.java:
```diff
@@ -115,4 +118,11 @@ public int size() {
     public long[] rowCounts() {
         return rowCounts;
     }
+
+    /**
+     * This is only a hint for a Session how to show the response for a user.
+     */
+    public boolean failFast() {
+        return failFast;
+    }
```
Makes sense. I got a comment on the prev iteration that used flag directly

https://crate.slack.com/archives/CBPLN6WJK/p1710236016595619

>I don't think the session should have insert specific knowledge like https://github.com/crate/crate/pull/15582/files#diff-74444ce9284113879dbc6e77596723633b15f43e3718b14fe22a4cbb3bf9e7d0L710-L715

I think point was to change components that Session works “as is” but as you see in the PR, we have to tweak `emitRowCountsToResultReceivers` a bit (specifically, call `ResultReciever.fail` for one bulk), so Session change for handling bulkResponse is inevitable.

I will adjust `emitRowCountsToResultReceivers` to get the flag value directly from Session.

BaurzhanSakhariev replied:
Done in https://github.com/crate/crate/pull/17689/commits/63077a9212b4dda69c2504e93457daa5bc50e686

BaurzhanSakhariev replied:
if I add `exceptionally` handling, I think I can revert all changes in `emitRowCountsToResultReceivers` and not pass the flag even here, checking

mfussenegger commented on server/src/main/java/io/crate/execution/engine/pipeline/ProjectionToProjectorVisitor.java:
```diff
@@ -559,7 +559,7 @@ public Projector visitUpdateProjection(final UpdateProjection projection, Contex
             ShardingUpsertExecutor.BULK_REQUEST_TIMEOUT_SETTING.get(settings),
             ShardUpsertRequest.DuplicateKeyAction.UPDATE_OR_FAIL,
-            true,
+            !context.txnCtx.sessionSettings().dmlFailFast(), // continueOnErrors,
```
Only in editors with inlay-hints enabled (which I find way too noisy to keep on). I think using comments for this (or extra local variables, e.g. `boolean continueOnErrors = !context.txnCtx.sessionSettings().dmlFailFast()` is fine to do.

BaurzhanSakhariev commented on server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:
```diff
@@ -2065,4 +2064,52 @@ public void test_can_write_ignored_object_with_array_column_with_mixed_types() t
+    }
+
+    @Test
+    public void test_bulk_insert_from_values_fail_fast() throws Exception {
+        execute("create table t (a int primary key)");
+
+        try (var session = sqlExecutor.newSession()) {
+            session.sessionSettings().dmlFailFast(true);
+            Object[][] bulkArgs = new Object[][]{
+                new Object[]{1},
+                new Object[]{1}
+            };
+            assertSQLError(() -> executeBulk("insert into t (a) values (?)", bulkArgs, session))
+                .hasPGError(UNIQUE_VIOLATION)
+                .hasHTTPError(CONFLICT, 4091)
+                .hasMessageContaining("A document with the same primary key exists already");
+        }
+
+        // First error encountered is reflected in jobs_log.
+        var response = execute("""
+                SELECT error FROM sys.jobs_log WHERE stmt LIKE 'insert into t (a) values%'
+                """);
+        assertThat((String) response.rows()[0][0]).contains("A document with the same primary key exists already");
+    }
```
This covers code being discussed in https://github.com/crate/crate/pull/17689#discussion_r2022459601

It hits different code path, if I revert `whenComplete((response, t) `handing in InsertFromValues this tests gets stuck and `test_insert_from_values_fail` passes.

BaurzhanSakhariev replied:
Can remove it now as I'm going to revert that special handling

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/indexing/IndexWriterProjector.java:
```diff
@@ -110,7 +110,7 @@ public IndexWriterProjector(ClusterService clusterService,
             ShardingUpsertExecutor.BULK_REQUEST_TIMEOUT_SETTING.get(settings),
             overwriteDuplicates ? DuplicateKeyAction.OVERWRITE : DuplicateKeyAction.UPDATE_OR_FAIL,
-            true,
+            true, // dml_fail_fast setting is never applied, COPY FROM has own fail_fast and RETURN SUMMARY
```
It's used in COPY FROM, I treated it as something  similar to INSERT ie roughly also DML in a way, but non insisting on keeping this comment

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java:
```diff
@@ -367,8 +367,14 @@ private void countdown() {
         }
 
+        /**
+         * @param failure is not NULL only when continueOnError is false.
+         * Hence, can use interrupt not only for KILL requests
+         * but also to reflect continueOnError derived from the dml_fail_fast setting.
+         */
         private void maybeSetInterrupt(@Nullable Exception failure) {
-            if (failure instanceof InterruptedException) {
+            if (failure != null) {
+                // No need to call prepareForClientTransmission, it's done later.
                 interrupt.set(failure);
             }
         }
```
Removed the method.

COPY FROM is not affected because we never flip the flag in the `IndexWriterProjector`.

And from the InterruptedException point of view (KILL) we didn't change anything here.

BaurzhanSakhariev commented on server/src/main/java/io/crate/planner/operators/InsertFromValues.java:
```diff
@@ -357,13 +362,20 @@ public CompletableFuture<BulkResponse> executeBulk(DependencyCarrier dependencie
+        AccessControl accessControl = dependencies
+            .nodeContext()
+            .roles()
+            .getAccessControl(sessionSettings.authenticatedUser(), sessionSettings.sessionUser());
```
Moved right before the usage.

kept `var sessionSettings = plannerContext.transactionContext().sessionSettings() `on it current place as it's used for computing continueOnError

BaurzhanSakhariev commented on server/src/main/java/io/crate/session/Session.java:
```diff
@@ -763,19 +763,37 @@ private static void emitRowCountsToResultReceivers(UUID jobId,
+        } else {
+            jobsLogs.logExecutionEnd(jobId, null);
         }
-        jobsLogs.logExecutionEnd(jobId, null);
```
I will change my approach and revert this and `InsertFromValues` bulkExecution changes. Also related to UpdateById stopping early.

BaurzhanSakhariev commented:
Hi @seut, could you please check the last commit https://github.com/crate/crate/pull/17689/commits/20363eba3200bca0165f9a89604bb12c16419c45?

it addressed multiple comments and basically changed approach
1. No special handling in InsertFromValues, reverted to a master state, relates to https://github.com/crate/crate/pull/17689#discussion_r2022459601
2. reverted changes in `emitRowCountsToResultReceivers` to the master satte, so not only removed flag from `BulkResponse` but not even using it in the `Session`.
3. removed redundant `test_bulk_insert_from_values_fail_fast`, extracted sys.job_log check from it and added to another test that was lacking it.
4. added fail fast to `UpdateByID` and updated tests
5. Regarding `exceptionally`:
not sure if we need to fail or finish result recievers, 
both `toExec.get(i).resultReceiver().fail(t)` and `toExec.get(i).resultReceiver().allFinished()` pass tests.

6. Looks like we need to use `prepareForClientTransmission` for sys table messages, see `test_insert_from_values_fail`.

Thrown exception is transformed somewhere else to `A document with the same primary key exists already `but sys.jobs_log has not transformed entry `version conflict, document already exists`

proddata commented on docs/config/session.rst:
```diff
@@ -276,4 +276,14 @@ consumption for the particular query.
   cr> SET search_path TO DEFAULT;
   SET OK, 0 rows affected (... sec)
 
+.. _conf-session-fail-writes:
+
+**allow_fail_on_partial_writes**
+  | *Default:* ``false``
+  | *Modifiable:* ``yes``
+   This setting enables partial failures of ``INSERT`` and ``UPDATE``
+   statements. If enabled, the first error encountered gets propagated and the
+   operation is terminated. The number of already written records is not
+   reported.
+
```
I think the name of the setting is misleading. Partial writes and partial failures can already occur regardless of this setting. The setting doesn’t enable partial failures - they are possible by design in CrateDB.Instead, it controls whether a write operation should abort immediately on the first row-level error, returning an error to the client, or continue silently, ignoring errors and returning the number of successfully processed rows.

It should also be clearly stated that even if an error is returned, some rows may already have been written.


`abort_on_first_row_error`, `fail_fast_on_row_error` or something like `write_fail_fast` should align better with the expected behaviour.

Also this seems to be similiar to `fail_fast` option in `COPY FROM` which also allows partial writes and is described as:

> A boolean value indicating if the COPY FROM operation should abort early after an error. This is best effort and due to the distributed execution, it may continue processing some records before it aborts.

---

Also, if this setting affects HTTP bulk requests - which, in my opinion, it **should not** - this must be clearly documented both here and in the section describing bulk request error behavior. Bulk requests are often used specifically to handle a few expected row-level conflicts (e.g., duplicates), which can be easily identified and handled gracefully. If this setting changes that behavior, its usage - especially setting sessions defaults - need to be carefully evaluated, as it may break common usage patterns with bulk requests.

seut replied:
> Also, if this setting affects HTTP bulk requests - which, in my opinion, it should not - this must be clearly documented both here and in the section describing bulk request error behavior.

I don't understand why this new session setting should not be taken into account when using the HTTP client API. If the bulk requests should not fail-fast, the user don't have to enable it on the used session.

proddata replied:
> I don't understand why this new session setting should not be taken into account when using the HTTP client API. 

Bulk requests already have a specific error reporting mechanism that allows identifying problematic records or executions by reporting both failed and successful entries. Additionally, due to the typical size limitations of bulk requests, fast-failing becomes less relevant or even redundant.

> If the bulk requests should not fail-fast, the user don't have to enable it on the used session.
It’s far from trivial with the HTTP endpoint, as the session is tied to the underlying TCP connection, which cannot be reliably controlled. Additionally, since there has been discussion about changing the default behavior for this change - or also when  allowing user- or role-specific defaults - this adds further complexity.

BaurzhanSakhariev replied:
On naming:

there was a [comment](https://github.com/crate/crate/pull/17689#discussion_r2022454841) to highlight a partial write. 

If we want to emphasize non-silent behavior caused by the setting and still highlight "partial write" aspect for those users to whom it might not be obvious, I suggest to name it `fail_fast_on_partial_writes`. wdyt?

proddata replied:
Don’t partial writes happen regardless of the setting? From what I understand the setting only changes the error behaviour (ignored or raised immediatily after first ecountered error)

We could also consider ...
- ignore_errors_on_writes (default: true)
- skip_failing_rows (default: true)
- continue_on_write_error (default: true)

BaurzhanSakhariev replied:
`continue_on_write_error` sounds good. 

@mfussenegger because you suggested to rename - is it okay to name it like that?

BaurzhanSakhariev replied:
As discussed somewhere else, will rename to `insert_select_fail_fast` and reduce the scope to insert-from-select once I pass review on implementation.

seut commented on server/src/main/java/io/crate/execution/dml/BulkShardResponseListener.java:
```diff
@@ -22,6 +22,7 @@
 package io.crate.execution.dml;
 
 import java.util.concurrent.CompletableFuture;
+import java.util.function.BiConsumer;
```
Please revert all changes inside this file, they are not needed. Set the `continueOnError` at `UpdateById` based on `plannerContext.transactionContext().sessionSettings().allowFailOnPartialWrites()` instead.

seut commented on server/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java:
```diff
@@ -346,7 +345,10 @@ private class ShardResponseActionListener implements ActionListener<ShardRespons
             resultAccumulator.accept(upsertResults, shardResponse, rowSourceInfos);
-            maybeSetInterrupt(shardResponse.failure());
+            var failure = shardResponse.failure();
+            if (failure != null) {
+                interrupt.set(failure);
+            }
```
```suggestion
            interrupt.set(failure);
```

seut reviewed (commented):
> Looks like we need to use prepareForClientTransmission for sys table messages, see test_insert_from_values_fail.
> 
> Thrown exception is transformed somewhere else to A document with the same primary key exists already but sys.jobs_log has not transformed entry version conflict, document already exists

As said, this seems to be a general issue and should be probably be fixed for all logs with errors.

BaurzhanSakhariev commented:
>As said, this seems to be a general issue and should be probably be fixed for all logs with errors.

Thanks, I overlooked it, will follow up later (independently on this PR)

seut commented on server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:
```diff
@@ -2065,4 +2064,40 @@ public void test_can_write_ignored_object_with_array_column_with_mixed_types() t
+
+    @Test
+    public void test_insert_from_values_fail() throws Exception {
+        execute("create table t (a int primary key, b int)");
```
```suggestion
        execute("create table t (a int primary key, b int) clustered into 1 shards");
```

seut commented on server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:
```diff
@@ -2065,4 +2064,40 @@ public void test_can_write_ignored_object_with_array_column_with_mixed_types() t
+        execute("create table t (a int primary key, b int)");
+        try (var session = sqlExecutor.newSession()) {
+            session.sessionSettings().allowFailOnPartialWrites(true);
+            assertSQLError(() -> execute("insert into t (a,b) values (1, 1),  (1, 2)", session))
```
```suggestion
            assertSQLError(() -> execute("insert into t (a,b) values (1, 1),  (1, 2), (2, 2)", session))
```

seut commented on server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:
```diff
@@ -2065,4 +2064,40 @@ public void test_can_write_ignored_object_with_array_column_with_mixed_types() t
+        var response = execute("""
+                SELECT error FROM sys.jobs_log WHERE stmt LIKE 'insert into t (a,b) values %'
+                """);
+        assertThat((String) response.rows()[0][0]).contains("version conflict, document already exists");
```
with the changes above, please assert that row (2, 2) was not written.

seut commented on server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:
```diff
@@ -2065,4 +2064,40 @@ public void test_can_write_ignored_object_with_array_column_with_mixed_types() t
+
+    @Test
+    public void test_insert_from_select_fail_fast() throws Exception {
+        execute("create table t (a int NOT NULL, b int)");
```
```suggestion
        execute("create table t (a int NOT NULL) clustered into 1 shards");
```

seut commented on server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:
```diff
@@ -2065,4 +2064,40 @@ public void test_can_write_ignored_object_with_array_column_with_mixed_types() t
+        execute("create table t (a int NOT NULL, b int)");
+        try (var session = sqlExecutor.newSession()) {
+            session.sessionSettings().allowFailOnPartialWrites(true);
+            assertSQLError(() -> execute("insert into t (a,b) select NULL, 1", session))
```
```suggestion
            assertSQLError(() -> execute("insert into t (a) select unnest([NULL, 1])", session))
```

seut commented on server/src/test/java/io/crate/integrationtests/InsertIntoIntegrationTest.java:
```diff
@@ -2065,4 +2064,40 @@ public void test_can_write_ignored_object_with_array_column_with_mixed_types() t
+                .hasMessageContaining("\"a\" must not be null");
+
+        }
+
```
same here, with the above changes, please assert that row (1) was not written

seut commented on server/src/test/java/io/crate/integrationtests/UpdateIntegrationTest.java:
```diff
@@ -1134,4 +1134,54 @@ public void test_update_continues_on_error() {
+        try (var session = sqlExecutor.newSession()) {
+            session.sessionSettings().allowFailOnPartialWrites(true);
+            Object[][] bulkArgs = new Object[][]{
+                new Object[]{2},
+                new Object[]{3}
```
Please use a valid row after a invalid one to verify that the valid one isn't written.
```suggestion
                new Object[]{2},
                new Object[]{1}
```

seut commented on server/src/testFixtures/java/io/crate/testing/SQLTransportExecutor.java:
```diff
@@ -310,15 +310,15 @@ private void executeBulk(String stmt,
+        if (bulkArgs != null && bulkArgs.length == 0) {
+            listener.onResponse(new BulkResponse(0));
+            return;
+        }
```
Is this change needed? If not, maybe revert?

seut reviewed (approved):
Code wise this looks good now, thx!
But please wait until the setting naming discussion is solved.

jeeminso commented:
I think we should utilize `earlyTermination` condition, If we flip https://github.com/crate/crate/blob/6e913b36c2674daca8e03555fd4da911ad2cb934/server/src/main/java/io/crate/execution/engine/indexing/ColumnIndexWriterProjector.java#L150
to `true`:
```
cr> create table t (a int check(a != 10));
CREATE OK, 1 row affected (2.444 sec)
cr> insert into t select * from generate_series(1, 100000);
JobKilledException[Job killed. 
[URI: null, ERRORS: {}]]
cr> select count(*) from t;
+----------+
| count(*) |
+----------+
|    55021 |
+----------+
SELECT 1 row in set (0.069 sec)
```
we end up with 50% inserted but this PR seems to insert over 90%:
```
cr> create table t (a int check(a != 10));
CREATE OK, 1 row affected (2.365 sec)
cr> set allow_fail_on_partial_writes = true;
SET OK, 0 rows affected (0.008 sec)
cr> insert into t select * from generate_series(1, 100000);
INSERT OK, 97571 rows affected (0.688 sec)
```
Could you double check @BaurzhanSakhariev?

BaurzhanSakhariev commented:
> I think we should utilize `earlyTermination` condition, If we flip
> 

While checking this, I found a bug - error is not thrown sometimes at all for `insert from select`

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java:
```diff
@@ -116,11 +117,12 @@ public CompletableFuture<R> consumeIteratorAndExecute() {
+        if (earlyTerminationCondition.apply(result, failure)) {
+            setResult(null, resultsToFailure.apply(result, failure));
+            return;
+        }
+
         if (result != null) {
-            if (earlyTerminationCondition.test(result)) {
-                setResult(null, resultsToFailure.apply(result));
-                return;
-            }
             resultRef.accumulateAndGet(result, combiner);
```
for fail fast on insert we use short-circuit mechanism via propagating `shardResponse.failure`, so `ShardResponseActionListener` can complete with an exception 
and `continueConsumptionOrFinish` receives args (null, throwable). 

Hence, moved the check out of the `if (result != null)`


BTW, we have a code branch on master, that hits `(null, maybe-not-null)`
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java#L197

I wonder if it means that `COPY FROM` with fail_fast = true can skip fail-fast logic for that scenario.

jeeminso replied:
Looks like `shardResponse.failure` is passed via `UpsertResults` not via the `failure` parameter of `continueConsumptionOrFinish`. So I do not think we need to move the code block out of the `if` stmt.


> BTW, we have a code branch on master, that hits (null, maybe-not-null)
https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java#L197
I wonder if it means that COPY FROM with fail_fast = true can skip fail-fast logic for that scenario

The scenario here is when an exception is thrown during loading next batch? I think you are right, that we should terminate early here. (just wondering why we never terminated in such case before, even before earlyTermination was added)

BaurzhanSakhariev replied:
> So I do not think we need to move the code block out of the if stmt.

it won't work if I do it, test will again insert almost 60000 records. 

>Looks like shardResponse.failure is passed via UpsertResults

No, see comment below

>I initially started with modifying RowCountCollector.processShardResponse to account for that standalone failure but, as mentioned above (https://github.com/crate/crate/pull/17689#discussion_r2035155167), with "short-circuit error" we complete future exceptionally and hence continueConsumptionOrFinish receives null-result and not-null throwable on failure.

seut reviewed (commented):
@BaurzhanSakhariev
Instead of implementing failure handling at the earlyTerminationCondition, I think we should stop consuming if the failure is set in general.
So stop consuming by setting consumptionFinished = true when failure != null. I wonder why we didn't do this and why this didn't lead to issues yet.

BaurzhanSakhariev commented:
> So stop consuming by setting consumptionFinished = true when failure != null. 


`failure != null` here happens when `ShardResponseActionListener.countdown` completes with exception. I think it means:

1. For this feature we can indeed do that
2.  On master, failure != null can be set only on KILL. So I think we have an issue here, that I would phrase as 
     "COPY FROM /INSERT don't properly react to KILL, if inFlight>0, we will try to again consume iterator.
 3. For COPY FROM I'm not sure it will work -  afaik fail_fast in COPY FROM doesn't set shardResponse.failure as it doesn't utilize continueOnError. INstead, it collects "silent failures" and eventually ShardResponseActionListener.countdown completes non-exceptionally with a non-null result (that however holds failures)
 

I will double check 2 and 3.

BaurzhanSakhariev commented on server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java:
```diff
@@ -116,11 +117,14 @@ public CompletableFuture<R> consumeIteratorAndExecute() {
+        if (failure != null) {
+            consumptionFinished = true;
+            // We night have inFlight > 0, so set the result and short circuit.
+            setResult(null, failure);
```
Actually, test is failing again with 57479L rows inserted - we should check the condition also when result is null (and failure is not null)

BaurzhanSakhariev replied:
Thanks, I reverted everything to the master state and only flipping consumptionFinished when failure is not null.

I have 2 questions:

1. We basically don't check fail fast/continueOnError flag - it makes sense, just wanted to clarify. 
 Is it irrelevant to check the flag because failure means we "short-circuited" due to KILL or setting being enabled so need to stop operation anyway?
2. I got test failing with 57479L records inserted (instead of <50000) when I only flip `consumptionFinished` flag. Difference when we set the result and exit is that we immediately close the batch iterator. To my understanding, we don't consume the iterator but let ongoing operations to finish and closing batch iterator stops even those(?). I pushed a minor change https://github.com/crate/crate/pull/17689/commits/d5bf7a9831857aa74ffab0a8b10d08548e932742 and got test passing.

BaurzhanSakhariev commented:
Hi @seut, could you please take another look? 

This is not a draft as I reduced the scope to insert-from-select (as discussed in different channels) and renamed the setting

BaurzhanSakhariev commented:
retest this please

seut commented on server/src/main/java/io/crate/execution/engine/indexing/BatchIteratorBackpressureExecutor.java:
```diff
@@ -123,8 +123,10 @@ private void continueConsumptionOrFinish(@Nullable R result, Throwable failure)
+
         if (failure != null) {
-            failureRef.set(failure);
+            setResult(null, failure);
+            return;
```
```suggestion
            consumptionFinished = true;
```

seut replied:
> Thanks, I reverted everything to the master state and only flipping `consumptionFinished` when failure is not null.
> 
> I have 2 questions:
> 
>     1. We basically don't check fail fast/continueOnError flag - it makes sense, just wanted to clarify.
>        Is it irrelevant to check the flag because failure means we "short-circuited" due to KILL or setting being enabled so need to stop operation anyway?

The fail-fast/continueOnError flag is only indicated when the writer, the shard upsert action should stop in case of a failure. the failure is then set immediately on the response.
Such, the component consuming the response do not have to care why and when a failure was set, it just have to process the failure.

> 
>     2. I got test failing with 57479L records inserted (instead of <50000) when I only flip `consumptionFinished` flag. Difference when we set the result and exit is that we immediately close the batch iterator. To my understanding, we don't consume the iterator but let ongoing operations to finish and closing batch iterator stops even those(?). I pushed a minor change [d5bf7a9](https://github.com/crate/crate/commit/d5bf7a9831857aa74ffab0a8b10d08548e932742) and got test passing.

I wonder these observation is related to where the operation is run (single node vs multi node), because closing the batch iterator will not stop any component on the other node (the writer).
So I could not understand why closing the iterator should stop the writer.

But I think we still could stop immediately like you did as we're anyway not interested in anymore responses of possible ongoing operations. But as said, I don't think doing this or not should affect how much rows are written.

BaurzhanSakhariev replied:
I tried with a bigger number (100_000), and run test multiple times, there is quite some difference, it's not negligible.

- flip flag: 97479L rows inserted (consistently).

- setResult: fluctuates between 22000 and 32000. 

I debugged version with flip flag: after we encounter a failure and flip the flag execution still goes to another iteration of

```
 while (batchIterator.moveNext()) {
 
 execute(item) // item is ShardedRequests
 }
```

and setResult closes the iterator so it can't move further.

BaurzhanSakhariev replied:
Hi @seut wdyt about 

>handler still sending requests to writers

does it make sense?

itest timeouts with more data but I can run more experiments with an actual cluster and insert millions of rows and track the difference (maybe even try to track difference per node). Can add some logging when we flip the flag/stop the operation and check if other nodes are still getting ShardUpserRequest(s) after we saw a first failure log entry.

BaurzhanSakhariev replied:
Hi @seut wdyt about ^? 

Should I experiment with higher numbers?

seut reviewed (approved):
👍
